File: 2024/05/day_05.py
from typing import List, Dict, Set, Optional

class Solution:
    def __init__(self):
        self.rules:Dict[int, List[int]] = {}

    def check_ordering(self, ordering: List[int]) -> bool:
        for i in range(len(ordering)):
            previous_items = ordering[:i]
            intersect = [x for x in self.rules.get(ordering[i], []) if x in previous_items]
            if len(intersect) > 0:
                return False
        return True
    
    def reorder(self, ordering: List[int]) -> List[int]:
        new_output: List[int] = []
        for i in range(len(ordering)):
            previous_items = new_output[:i]
            intersect = set([x for x in self.rules.get(ordering[i], []) if x in previous_items])
            if len(intersect) > 0:
                # find the earliest index of any element in the index list
                index = min([new_output.index(x) for x in intersect])
                new_output.insert(index, ordering[i])
            else:
                new_output.append(ordering[i])
        return new_output
    
    def add_rule(self, key: int, val: int) -> List[int]:
        if key in self.rules:
            self.rules[key].append(val)
        else:
            self.rules[key] = [val]
        return self.rules[key]
        
    def do_all(self, input: str) -> int:
        total_sum = 0
        for line in input.splitlines():
            if "|" in line:
                k, v = [int(x) for x in line.split("|")]
                self.add_rule(k, v)
            if "," in line:
                ordering = [int(x) for x in line.split(",")]
                if self.check_ordering(ordering):
                    total_sum = total_sum + ordering[len(ordering)//2]
        return total_sum

    def do_all_reordering(self, input: str) -> int:
        total_sum = 0
        for line in input.splitlines():
            if "|" in line:
                k, v = [int(x) for x in line.split("|")]
                self.add_rule(k, v)
        
        for line in input.splitlines():
            if "," in line:
                ordering = [int(x) for x in line.split(",")]
                if not(self.check_ordering(ordering)):
                    new_ordering = self.reorder(ordering)
                    total_sum = total_sum + new_ordering[len(new_ordering)//2]
        return total_sum

    
def main(fn: str):
    with open(fn) as f:
        s = Solution()
        data = f.read()
        print(s.do_all(data))
        s2 = Solution()
        print(s2.do_all_reordering(data))
    
if __name__ == "__main__":
    main("input.txt")

File: 2024/05/day_05.jl
using Printf

mutable struct Solution

    rules::Dict{Int, Vector{Int}}
    flat_rules::Dict{Int, Vector{Int}}

    function Solution()
        new(Dict(), Dict())
    end
end


function check_ordering(solution::Solution, ordering::Vector{Int})::Bool
    for i in eachindex(ordering)
        previous_items = ordering[1:(i-1)]
        intersect = [x for x in get(solution.rules, ordering[i], []) if x in previous_items]
        if length(intersect) > 0
            return false
        end
    end
    return true 
end

function reorder(solution::Solution, ordering::Vector{Int})::Vector{Int}
    new_output::Vector{Int} = []
    for i in eachindex(ordering)
        intersect = [x for x in get(solution.rules, ordering[i], []) if x in new_output]
        if length(intersect) > 0
            indices = [findfirst(x->x==y, new_output) for y in intersect]
            new_index = minimum(i for i in indices if i !== nothing)
            insert!(new_output, new_index, ordering[i])
        else
            push!(new_output, ordering[i])
        end
    end
    return new_output
end

function add_rule(solution::Solution, key::Int, val::Int)::Vector{Int}
    if haskey(solution.rules, key)
        push!(solution.rules[key], val)
    else
        solution.rules[key] = [val]
    end
    return solution.rules[key]
end

function do_all(solution::Solution, lines::Vector{String})::Int
    total_sum = 0
    for line in lines
        if occursin("|", line)
            k, v = map(x->parse(Int, x), split(line, "|"))
            add_rule(solution, k, v)
        end
        if occursin(",", line)
            ordering = map(x->parse(Int, x), split(line, ","))
            if check_ordering(solution, ordering)
                middle_index = ceil(Int, length(ordering) / 2)
                total_sum = total_sum + ordering[middle_index]
            end
        end
    end
    return total_sum
end

function do_all_reordering(solution::Solution, lines::Vector{String})::Int
    total_sum = 0
    for line in lines
        if occursin("|", line)
            k, v = map(x->parse(Int, x), split(line, "|"))
            add_rule(solution, k, v)
        end
    end

    for line in lines
        if occursin(",", line)
            ordering = map(x->parse(Int, x), split(line, ","))
            if !check_ordering(solution, ordering)
                new_ordering = reorder(solution, ordering)
                middle_index = ceil(Int, length(new_ordering) / 2)
                total_sum = total_sum + new_ordering[middle_index]
            end
        end
    end
    return total_sum
end

function read_data(fn::String)::Vector{String}
    return readlines(fn)
end

if abspath(PROGRAM_FILE) == @__FILE__
    data = read_data("input.txt")

    s = Solution()
    println(do_all(s, data))

    s = Solution()
    println(do_all_reordering(s, data))
end
